core/gotemplate: Add engine tests, unlock mutex in case of error

carstendietrich opened the pull request:
Add tests for the engine
Fix mutex staying locked in case of error which can lead to a deadlock
Fix template function `map` causing out of bounds error if called with not enough arguments

thebod commented on core/gotemplate/engine.go:
```diff
@@ -64,14 +64,14 @@ func (e *engine) Render(ctx context.Context, name string, data interface{}) (io.
 
-	lock.Lock()
 	if e.debug || e.templates == nil {
+		lock.Lock()
```
This will lead to a race condition when e.templates is nil, eventually triggering e.loadTemplates multiple times. Move lock.Lock() up, and call lock.Unlock() both in error and non-error case.

thebod commented on core/gotemplate/engine_test.go:
```diff
@@ -0,0 +1,183 @@
+	return funcs
+}
+
+type CustomTemplateFunc struct{}
```
Don't export

thebod commented on core/gotemplate/engine_test.go:
```diff
@@ -0,0 +1,183 @@
+	}
+	tests := []struct {
+		name    string
+		fields  fields
```
It is not clear what `fields` and `args` mean, maybe rename fields to engineConfig and args to renderArgs?

thebod commented:
Unittests in 1.11 are failing, despite restart

carstendietrich commented:
Hey @bastianccm,

fixed the mentioned stuff and switched to go-cmp instead of reflect.DeeplEqual. Tests in Go 1.11.x are now running as well. Please have a look.

ty
